pkg/k8s/rule_translate.go

```diff
@@ -0,0 +1,180 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+var _ policy.Translator = RuleTranslator{}
+
+// RuleTranslator implements pkg/policy.Translator interface
+// Translate populates/depopulates given rule with ToCIDR rules
+// Based on provided service/endpoint
+type RuleTranslator struct {
+	Service  types.K8sServiceNamespace
+	Endpoint types.K8sServiceEndpoint
+	Revert   bool
+}
+
+// Translate calls TranslateEgress on all r.Egress rules
+func (k RuleTranslator) Translate(r *api.Rule) error {
+	for egressIndex := range r.Egress {
+		err := k.TranslateEgress(&r.Egress[egressIndex])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// TranslateEgress populates/depopulates egress rules with ToCIDR entries based
+// on toService entries
+func (k RuleTranslator) TranslateEgress(r *api.EgressRule) error {
+	err := k.depopulateEgress(r)
+	if err != nil {
+		return err
+	}
+	if !k.Revert {
+		err := k.populateEgress(r)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k RuleTranslator) populateEgress(r *api.EgressRule) error {
+	for _, service := range r.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == api.K8sServiceNamespace(k.Service) {
+			if err := generateToCidrFromEndpoint(r, k.Endpoint); err != nil {
+				return err
+			}
+			// TODO: generateToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+func (k RuleTranslator) depopulateEgress(r *api.EgressRule) error {
+	for _, service := range r.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == api.K8sServiceNamespace(k.Service) {
+			if err := deleteToCidrFromEndpoint(r, k.Endpoint); err != nil {
+				return err
+			}
+			// TODO: generateToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+// generateToCidrFromEndpoint takes an egress rule and populates it with
+// ToCIDR rules based on provided endpoint object
+func generateToCidrFromEndpoint(
+	egress *api.EgressRule, endpoint types.K8sServiceEndpoint) error {
+
+	// This will generate one-address CIDRs consisting of endpoint backend ip
+	mask := net.CIDRMask(128, 128)
+	for ip := range endpoint.BEIPs {
+		epIP := net.ParseIP(ip)
+		if epIP == nil {
+			return fmt.Errorf("Unable to parse ip: %s", ip)
+		}
+
+		found := false
+		for _, c := range egress.ToCIDR {
+			_, cidr, err := net.ParseCIDR(string(c))
+			if err != nil {
+				return err
+			}
+			if cidr.Contains(epIP) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			cidr := net.IPNet{IP: epIP.Mask(mask), Mask: mask}
+			egress.ToCIDR = append(egress.ToCIDR, api.CIDR(cidr.String()))
```

Comment:
In general, a basic principle in any REST API is that entities must never be modified except as a result of a client requests.
I.e. if I import a policy with `cilium policy import`, I should always get the exact same policy from `cilium policy get` afterwards. User input should never be modified.
You're violating that basic principle.

Reply:
This is true in context of using `cilium` as a user-facing client. This code is kubernetes-specific and alters what seems like an internal representation of k8s CiliumNetworkPolicy Custom Resource Definition. User which would be affected by this code would use `kubectl create -f policy.yaml`, and see the same objects when he retrieves his policy from kubernetes with `kubectl get cnp -o yaml`, as this code doesn't manipulate k8s objects.

Please let me know if this is acceptable for you.

Reply:
Got it. I didnt understand that. Thanks!

Reply:
What's preventing a user from defining a rule with a non-empty `ToServices` rule and load it with `cilium policy import`? What is that supposed not to work?
Has that been tested? I don't see any integration / component tests in this PR.